src/wagtail_personalisation/wagtail_hooks.py

```diff
@@ -250,3 +254,49 @@ def add_personalisation_summary_panels(request, items):
     items.append(SegmentSummaryPanel(request))
     items.append(PersonalisedPagesSummaryPanel(request))
     items.append(VariantPagesSummaryPanel(request))
+
+
+@hooks.register('before_delete_page')
+def delete_related_variants(request, page):
+    if not isinstance(page, models.PersonalisablePageMixin) \
+            or not page.personalisation_metadata.is_canonical:
+        return
+    # Get a list of related personalisation metadata for all the related
+    # variants.
+    variants_metadata = (
+        page.personalisation_metadata.variants_metadata
+                                     .select_related('variant')
+    )
+    next_url = get_valid_next_url_from_request(request)
+
+    if request.method == 'POST':
+        parent_id = page.get_parent().id
+        variants_metadata = variants_metadata.select_related('variant')
+        with transaction.atomic():
+            for metadata in variants_metadata.iterator():
+                # Call delete() on objects to trigger any signals or hooks.
+                metadata.variant.delete()
+            # Delete the page's main variant and the page itself.
+            page.personalisation_metadata.delete()
+            page.delete()
+        msg = _("Page '{0}' and its variants deleted.")
+        messages.success(
+            request,
+            msg.format(page.get_admin_display_title())
+        )
+
```

Comment:
It would be a good idea to call the `after_delete_page` hooks here, in case there are other Wagtail add-ons that rely on it - it looks like that's the only bit from Wagtail's own `delete` method which isn't represented here.

(Having to replicate the `delete` view here definitely isn't ideal, and we should consider extending Wagtail's hooks to do whatever we need to avoid this - perhaps adding a transaction block around the deletion and a new hook that gets called inside that transaction.)

Other than that, looks good to me!

Reply:
PR to Wagtail - wagtail/wagtail#4673